agents/craftassist/craftassist_agent.py

```diff
@@ -225,17 +228,38 @@ def init_controller(self):
 
     def perceive(self, force=False):
         """Whenever some blocks are changed, that area will be put into a 
-        buffer which will be force-perceived by the agent in the next step
+        buffer which will be force-perceived by the agent in the next step.
 
         Here the agent first clusters areas that are overlapping on the buffer,
-        then run through all perception modules to perceive
-        and finally clear the buffer when perception is done.
+        then run through all perception modules to perceive and finally clears the
+        buffer when perception is done.
+
+        The agent sends all perception updates to memory in order for them to
+        update the memory state.
         """
         self.areas_to_perceive = cluster_areas(self.areas_to_perceive)
+        # 1. perceive from NLU parser
         super().perceive()
+        # 2. perceive from low_level perception module
+        perception_output = self.perception_modules["low_level"].perceive()
+        self.areas_to_perceive = self.memory.update_with_lowlevel_perception_input(
+            perception_output, self.areas_to_perceive)
+        # 3. perceive from heuristic perception module
+        heuristic_perception_output = {}
+        if force or not self.agent.memory.task_stack_peek():
+            # perceive from heuristic perception module
+            heuristic_perception_output = self.perception_modules["heuristic"].perceive()
+            self.memory.update_with_heuristic_perception_input(heuristic_perception_output)
```

Comment:
can this just be a self.memory.update(dict_with_all_changes), where dict_with_all_changes = {"blocks" : [...], "triples": [...], ...}

Reply:
yes, one method in memory that takes the entire dict of changes as input.   that method in memory can just run each of those separately... but from the agent side imo its better if there is a single method from memory that does all the updates

Reply:
The reason I split them into three sequential steps is because: `heuristic_perception`'s `perceive` works on top of updated `areas_to_perceive` from `low_lelvel`(https://github.com/facebookresearch/fairo/blob/8981edf1112375b37ce60d4b1db439bfbdfa3e1b/droidlet/perception/craftassist/heuristic_perception.py#L535). The update to agent's `areas_to_perceive` now happens inside the memory update after `low_level_perception`(https://github.com/facebookresearch/fairo/blob/ca_perception_refactor/droidlet/memory/craftassist/mc_memory.py#L270)

Reply:
yeah, I think its fine to split into several steps; but put imo its better if those steps are wrapped in a generalized memory.update(data).  if that update() needs to be run twice thats fine with me.  if we start needing to do clever things to deal with performance we can reconsider or refactor, but I think its better to try to keep the .update() interface uniform and not call the individual methods outside of memory

Reply:
Right yes, that's what I meant. I will update the function to be one big update (and updates will happen based on the changeset that comes in) but I'll just have to call `update` twice here - interleaving between perception calls